glean-core/src/upload/queue.rs

```diff
@@ -0,0 +1,138 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Pending pings queue.
+
+use std::collections::{HashSet, VecDeque};
+use std::sync::RwLock;
+
+use log;
+use uuid::Uuid;
+
+/// A queue like entity for enqueueing pending pings.
+///
+/// On dequeueing the pings are added to a `processing` set.
+/// They can't be added to the queue if they are being processed.
+pub struct PingsQueue {
+    /// A FIFO queue storing the UUID of the pending pings.
+    ///
+    /// These have not yet been sent for uploading.
+    queue: RwLock<VecDeque<Uuid>>,
+    /// A set storing the UUID of the pings being processed.
+    ///
+    /// These will have been sent for upload, but not yet
+    /// received confirmation that upload was successful.
+    processing: RwLock<HashSet<Uuid>>,
+}
+
+impl PingsQueue {
+    /// Create an empty PingsQueue.
+    pub fn new() -> Self {
+        Self {
+            queue: RwLock::new(VecDeque::new()),
+            processing: RwLock::new(HashSet::new()),
+        }
+    }
+
+    /// Enqueue a ping
+    ///
+    /// Pings that are in the processing set will be ignored
+    pub fn enqueue(&self, uuid: Uuid) {
+        let processing = self
```

Comment:
When would we happen to enqueue a ping that's already being processed?

Reply:
In case we process the directory while some request is still being processed.

Reply:
That would not happen if we were to scan only once at startup, right?

Reply:
Right.